Propagate kubeadm dual-stack feature-gate to all k8s components

Arvinderpal opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR propagates the kubeadm feature-gate, if set in ClusterConfiguration, to the all individual kubernetes components on the control-plane node.

```
kind: ClusterConfiguration
featureGates:
  IPv6DualStack: true
```

For worker nodes, users will have to specify the dual-stack feature-gate to kubelet using the nodeRegistration.kubeletExtraArgs option as part of their join config. Alternatively, they can use KUBELET_EXTRA_ARGS as well.

**Which issue(s) this PR fixes**:

This addresses one of the TODO items in the dual-stack kubeadm tracker ticket: https://github.com/kubernetes/kubeadm/issues/1612
It's a follow up to the initial kubeadm dual-stack feature-gate PR: https://github.com/kubernetes/kubernetes/pull/80145

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
In order to enable dual-stack support within kubeadm and kubernetes components, as part of the init config file, the user should set feature-gate IPv6DualStack=true in the ClusterConfiguration. Additionally, for each worker node, the user should set the feature-gate for kubelet using either nodeRegistration.kubeletExtraArgs or  KUBELET_EXTRA_ARGS.
```

Arvinderpal commented:
@neolit123 @yastij @fabriziopandini @aojea 
As requested, this PR propagates the dual-stack feature-gate flag to all kubernetes components. 
I made this a separate PR from the comma separated list support for --cluster-cidr. I did not want to mix these two together.

neolit123 commented:
/assign
/priority important-longterm
/ok-to-test

neolit123 commented on cmd/kubeadm/app/util/config/common.go:
```diff
@@ -193,3 +193,36 @@ func MigrateOldConfig(oldConfig []byte) ([]byte, error) {
+			extraArgs = make(map[string]string)
+		}
+		if _, present := extraArgs["feature-gates"]; present && len(extraArgs["feature-gates"]) > 0 {
+			extraArgs["feature-gates"] = strings.TrimSuffix(extraArgs["feature-gates"], ",")
```
interesting, so this handles a stray `,` at the end left by the user?

Arvinderpal replied:
Yes, it will handle a single stray `,`

neolit123 commented on cmd/kubeadm/app/util/config/common_test.go:
```diff
@@ -396,3 +396,56 @@ func TestMigrateOldConfigFromFile(t *testing.T) {
+		{nil, "IPv6DualStack=true"},
+	}
+	for _, rt := range tests {
+		actual := appendComponentFeatureGate(rt.extraArgs, testFeatureName)
```
please use sub-tests with `.name`:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/version_test.go#L72
[2]

Arvinderpal replied:
Done!

neolit123 commented on cmd/kubeadm/app/util/config/common_test.go:
```diff
@@ -396,3 +396,56 @@ func TestMigrateOldConfigFromFile(t *testing.T) {
+		{nil, false},
+	}
+	for _, rt := range tests {
+		actual := isComponentFeatureGatePresent(rt.extraArgs, testFeatureName)
```
[2]

Arvinderpal replied:
Done!

neolit123 reviewed (commented):
thanks @Arvinderpal 
added minor comments, but overall this is looking good.

neolit123 commented:
/retest

Arvinderpal commented on cmd/kubeadm/app/util/config/initconfiguration.go:
```diff
@@ -102,6 +105,11 @@ func SetNodeRegistrationDynamicDefaults(cfg *kubeadmapi.NodeRegistrationOptions,
 		klog.V(1).Infof("detected and using CRI socket: %s", cfg.CRISocket)
 	}
 
+	if isDualStack {
```
That won't work since we pass in `cfg *kubeadmapi.NodeRegistrationOptions` and not `ClusterConfiguration`

neolit123 replied:
ah i see, so one solution is to make SetNodeRegistrationDynamicDefaults accept ClusterConfiguration (it think it used to be like that at some point) :thinking: 

@rosti WDYT?
/cc @rosti

rosti replied:
This is not the place to do the feature gate patching for the kubelet. The correct place is `buildKubeletArgMap`:
https://github.com/kubernetes/kubernetes/blob/fc4cbf743ac899df1d96f520951c6b24dcaa8494/cmd/kubeadm/app/phases/kubelet/flags.go#L78

Arvinderpal replied:
@rosti I see a problem with how `BuildArgumentListFromMap` works today -- it does not handle comma separated fields like "feature-gates", since it does a simple overwrite of default values with user specified values. For example, if we set IPv6DualStack=true in base, then that will be overwritten if the user has specified some additional feature-gates in `KubeletExtraArgs`. 

We have a couple of options:
1. Make `BuildArgumentListFromMap` capable of "merging" fields like feature-gates rather than simple key/value overwrite. We would do something like this:
```	
for k, v := range overrideArguments {
		if k == "feature-gates" && len(argsMap[k]) > 0{
			argsMap[k] += "," + v
		}else{
			argsMap[k] = v
		}
	}
```
2. Modify top-level funcs like `WriteKubeletDynamicEnvFile` so that they perform the merge on feature-gates.  This would require updating the `argList` list returned by `BuildArgumentListFromMap`. This is not ideal, IMO.

We could also just leave the code in the higher-level funcs as it is in the current PR. 
I'll wait for you feedback before proceeding. Thanks.

rosti replied:
Let's keep things simple for now and just allow the users to overwrite the whole of the `feature-gates` flags if they wish.
Either way, this is going to be documented as part of the feature gate documentation.
Another possibility is to modify `BuildArgumentListFromMap` to warn on overwrites.

Certainly, we should not do this in the dynamic defaulting code.

@fabriziopandini @neolit123 WDYT?

neolit123 replied:
i think that @rosti is correct here that dynamic defaults is not the best place.

> Let's keep things simple for now and just allow the users to overwrite the whole of the feature-gates flags if they wish.

makes sense to me.

passing the dual stack feature gate to the kubelet will be implicit (if the kubeadm FG is on). so if the user provides their own feature-gates for the kubelet extraArgs, this should overwrite the implicit gate but should also show a warning.

> Either way, this is going to be documented as part of the feature gate documentation.

we might add a sentence about overrides, yes.

Arvinderpal replied:
Done.

Arvinderpal commented on cmd/kubeadm/app/util/config/initconfiguration.go:
```diff
@@ -155,6 +163,21 @@ func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, advertiseAd
 	// Downcase SANs. Some domain names (like ELBs) have capitals in them.
 	LowercaseSANs(cfg.APIServer.CertSANs)
+
+	if isDualStack {
```
Yes, here I could use `ClusterConfiguration` since that is passed into the func; however, just to be consistent with the `SetNodeRegistrationDynamicDefaults` approach, I use `isDualStack`

Arvinderpal commented on cmd/kubeadm/app/util/config/initconfiguration.go:
```diff
@@ -155,6 +163,21 @@ func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, advertiseAd
 	LowercaseSANs(cfg.APIServer.CertSANs)
+
+	if isDualStack {
+		// If "IPv6DualStack=true" is not present, we append it. This code should be remvoved after dual-stack is GA.
```
Done.

Arvinderpal commented on cmd/kubeadm/app/util/config/initconfiguration.go:
```diff
@@ -155,6 +163,21 @@ func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, advertiseAd
+			cfg.ComponentConfigs.KubeProxy.FeatureGates = make(map[string]bool)
+		}
+		if _, present := cfg.ComponentConfigs.KubeProxy.FeatureGates[features.IPv6DualStack]; !present {
+			cfg.ComponentConfigs.KubeProxy.FeatureGates[features.IPv6DualStack] = true
```
Yeah, completely agree!

rosti replied:
Again, this is not the place to do this:
- For the control plane components, look at the `get*Command` functions in:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/controlplane/manifests.go
- For the Kube Proxy, see `DefaultKubeProxyConfiguration`:
https://github.com/kubernetes/kubernetes/blob/fc4cbf743ac899df1d96f520951c6b24dcaa8494/cmd/kubeadm/app/componentconfigs/defaults.go#L38

Arvinderpal replied:
Please see my comment above. The same applies here with `BuildArgumentListFromMap`

Arvinderpal reviewed (commented):
@neolit123 Thanks for the feedback. Please see my comments.

aojea commented:
/test pull-kubernetes-integration

rosti reviewed (changes requested):
Thanks @Arvinderpal !
However, we need the feature gate propagation to be done at a more lower code level, than in the config dynamic defaulting code.

Arvinderpal committed:
Kubeadm FG for dual-stack was introduced in #80145. This PR
progagates the kubeadm FG to the individual k8scomponents
on the control-plane node.

* Note: Users who want to join worker nodes to the cluster
will have to specify the dual-stack FG to kubelet using the
nodeRegistration.kubeletExtraArgs option as part of their
join config. Alternatively, they can use KUBELET_EXTRA_ARGS.

kubeadm FG: kubernetes/kubeadm#1612

Arvinderpal commented:
@rosti @neolit123 Thank you for the feedback. I have made the changes requested. 
It actually makes things a lot simpler now :)

neolit123 reviewed (commented):
thanks for the update
/lgtm
approve on @rosti

neolit123 commented:
looks like some build tests are failing.

yastij commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -184,6 +185,15 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
+	enabled, present := cfg.FeatureGates[features.IPv6DualStack]
+	if present && enabled {
+		defaultArguments["feature-gates"] = features.IPv6DualStack + "=true"
+	} else if present && !enabled {
```
why not just `else` ? (the remaining case is `!present`)

rosti replied:
[1] Ideally, we would like to cover the `present==true` case. So something like this should work:
```golang
if present {
        defaultArguments["feature-gates"] = fmt.Sprintf("%s=%t", features.IPv6DualStack, enabled)
}
```

Arvinderpal replied:
Done!

yastij commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -309,6 +328,15 @@ func getSchedulerCommand(cfg *kubeadmapi.ClusterConfiguration) []string {
+	enabled, present := cfg.FeatureGates[features.IPv6DualStack]
+	if present && enabled {
+		defaultArguments["feature-gates"] = features.IPv6DualStack + "=true"
+	} else if present && !enabled {
```
same

rosti replied:
[1]

yastij commented on cmd/kubeadm/app/phases/kubelet/flags.go:
```diff
@@ -120,6 +121,15 @@ func buildKubeletArgMap(opts kubeletFlagsOpts) map[string]string {
+	enabled, present := opts.featureGates[features.IPv6DualStack]
+	if present && enabled {
+		kubeletFlags["feature-gates"] = features.IPv6DualStack + "=true"
+	} else if present && !enabled {
```
same

rosti replied:
[1]

yastij reviewed (commented):
just some nits, also +1 on the previous feedback @rosti

rosti commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -295,6 +305,15 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration) []string
+	// TODO: The following code should be remvoved after dual-stack is GA.
+	// Note: The user still retains the ability to explicitly set feature-gates and that value will overwrite this base value.
+	enabled, present := cfg.FeatureGates[features.IPv6DualStack]
+	if present && enabled {
```
[1]

rosti reviewed (commented):
Thanks @Arvinderpal !
Leaving final lgtm to @yastij or @neolit123 
/approve

yastij commented:
/lgtm

Arvinderpal commented:
/retest
